packages/query/src/block-processor/transactions.ts

```diff
@@ -67,6 +84,32 @@ export class Transactions {
       }
     }
   }
+
+  // LPNFT asset IDs won't be known to the chain, so we need to pre-populate them in local
+  // registry based on transaction contents.
+  // Necessary to save LPNFT denoms for 'closed' 'withdrawn' and 'claimed' as soon as we detect a positionOpen action
+  async storeLpNft(tx: Transaction) {
+    if (!tx.body?.actions) {
+      return;
+    }
+    for (const action of tx.body.actions) {
+      if (action.action.case === 'positionOpen' && action.action.value.position !== undefined) {
+        const position = action.action.value.position;
+
+        const opened = new PositionState({ state: PositionState_PositionStateEnum.OPENED });
+        await this.indexedDb.saveAssetsMetadata(this.viewServer.getLpNftDenom(position, opened));
+
+        const closed = new PositionState({ state: PositionState_PositionStateEnum.CLOSED });
+        await this.indexedDb.saveAssetsMetadata(this.viewServer.getLpNftDenom(position, closed));
+
+        const withdrawn = new PositionState({ state: PositionState_PositionStateEnum.WITHDRAWN });
+        await this.indexedDb.saveAssetsMetadata(this.viewServer.getLpNftDenom(position, withdrawn));
+
+        const claimed = new PositionState({ state: PositionState_PositionStateEnum.CLAIMED });
+        await this.indexedDb.saveAssetsMetadata(this.viewServer.getLpNftDenom(position, claimed));
+      }
+    }
+  }
```

Comment:
What do you think about this as an alernative?

```typescript
    for (const a of tx.body.actions) {
      if (a.action.case !== 'positionOpen' || a.action.value.position === undefined) {
        continue;
      }

      const positionStates = [
        PositionState_PositionStateEnum.OPENED,
        PositionState_PositionStateEnum.CLOSED,
        PositionState_PositionStateEnum.WITHDRAWN,
        PositionState_PositionStateEnum.CLAIMED,
      ];

      for (const state of positionStates) {
        const positionState = new PositionState({ state });
        const metadata = this.viewServer.getLpNftDenom(a.action.value.position, positionState);
        await this.indexedDb.saveAssetsMetadata(metadata);
      }
```

Reply:
Yes, it makes sense to avoid code duplication